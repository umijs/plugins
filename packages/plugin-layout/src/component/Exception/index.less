.umi-plugin-layout-exception-container {
  min-width: 750px;
  height: 624px;
  background-color: #fff;
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 24px;

  .umi-plugin-layout-exception-content {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: row;

    .umi-plugin-layout-exception-rightContent {
      margin-left: 120px;

      :global(h1.ant-typography) {
        font-size: 56px;
        line-height: 56px;
        margin-bottom: 0;
      }

      .umi-plugin-layout-exception-desc {
        margin: 8px 0 24px 0;
        font-size: 16px;
        color: rgba(0, 0, 0, 0.45);
        max-width: 350px;
      }
    }
  }
}
