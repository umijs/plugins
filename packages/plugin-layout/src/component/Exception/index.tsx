import React from 'react';
import { Result, Button } from 'antd';
import { history } from 'umi';
import { IRouteLayoutConfig } from '../../types/interface.d';

function backToHome() {
  history.push('/');
}

const Exception404 = () => (
  <Result
    status="404"
    title="404"
    subTitle="抱歉，你访问的页面不存在"
    extra={
      <Button type="primary" onClick={backToHome}>
        返回首页
      </Button>
    }
  />
);

const Exception403 = () => (
  <Result
    status="403"
    title="403"
    subTitle="抱歉，你无权访问该页面"
    extra={
      <Button type="primary" onClick={backToHome}>
        返回首页
      </Button>
    }
  />
);

const WithExceptionOpChildren: React.FC<{
  currentPathConfig?: IRouteLayoutConfig;
  children: any;
  noFound: React.ReactNode;
  unAccessible: React.ReactNode;
}> = props => {
  const { children, currentPathConfig } = props;
  // 404 现在应该很少会发生
  if (!currentPathConfig) {
    return props.noFound || <Exception404 />;
  }
  /**
   * 这里是没有权限的意思
   */
  if (currentPathConfig.unAccessible || currentPathConfig.unaccessible) {
    return props.unAccessible || <Exception403 />;
  }
  return children;
};

export { Exception404, Exception403, WithExceptionOpChildren };
