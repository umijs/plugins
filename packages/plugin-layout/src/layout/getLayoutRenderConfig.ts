const getLayoutRenderConfig = (currentPathConfig: {
  layout:
    | {
        hideMenu: boolean;
        hideNav: boolean;
        hideFooter: boolean;
      }
    | false;
  hideFooter: boolean;
}) => {
  const layoutRender: any = {};

  if (currentPathConfig?.hideFooter) {
    layoutRender.footerRender = false;
  }

  if (currentPathConfig?.layout == false) {
    layoutRender.pure = true;
    return layoutRender;
  }

  if (currentPathConfig?.layout?.hideMenu) {
    layoutRender.menuRender = false;
  }

  if (currentPathConfig?.layout?.hideFooter) {
    layoutRender.footerRender = false;
  }

  if (currentPathConfig?.layout?.hideNav) {
    layoutRender.headerRender = false;
  }

  return layoutRender;
};

export default getLayoutRenderConfig;
