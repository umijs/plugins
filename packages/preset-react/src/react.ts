import { IApi, utils } from 'umi';
import { join } from 'path';
import { readFileSync, existsSync } from 'fs';

const { semver } = utils;

export default (api: IApi) => {
  const isReact17 = () => {
    let react;
    try {
      react = require(require.resolve('react', { paths: [api.cwd] }));
    } catch (e) {}
    return (
      semver.valid(react?.version) &&
      semver.gte(react.version, '17.0.0-alpha.0')
    );
  };

  api.onStart(() => {
    const tsconfigPath = join(api.cwd, 'tsconfig.json');
    let tsconfigContent = '';
    try {
      if (existsSync(tsconfigPath)) {
        tsconfigContent = readFileSync(tsconfigPath, 'utf-8');
      }
    } catch (e) {}

    if (
      isReact17() &&
      tsconfigContent &&
      !tsconfigContent?.includes('react-jsx')
    ) {
      api.logger.warn(
        '[WARN] update `jsx: "react"` into `jsx: "react-jsx"` to suport the new JSX transform in React 17 in tsconfig.json',
      );
    }
  });

  // support react 17
  api.modifyBabelPresetOpts((opts) => {
    return {
      ...opts,
      reactRequire: !isReact17(),
      react: {
        ...(opts.react || {}),
        // support React 17 New Jsx syntax
        runtime: isReact17() ? 'automatic' : 'classic',
      },
    };
  });
};
