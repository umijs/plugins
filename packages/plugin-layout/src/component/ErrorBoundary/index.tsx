import * as React from 'react';
import { Result, Typography } from 'antd';
import { intl } from '../../utils/intl';
import { Exception500 } from '../Exception';

interface Error {
  componentStack?: string;
  error?: string;
  [key: string]: any;
}

interface IState {
  hasError?: boolean;
  error?: string;
  info?: {
    componentStack?: string;
  };
}

export interface IProps {
  /** 发生错误后的回调（可做一些错误日志上报，打点等） */
  onError?: (error: Error, info: any) => void;
}

const DefaultFallbackComponent = ({ componentStack, error }: Error) => (
  <Result
    status="error"
    title={intl({ id: 'layout.global.error.title' })}
    subTitle={error!.toString()}
  >
    <Typography.Paragraph>
      {intl({ id: 'layout.global.error.stack' })}：<pre>{componentStack}</pre>
    </Typography.Paragraph>
  </Result>
);

class ErrorBoundary extends React.Component<IProps, IState> {
  static defaultProps = {
    onError: null,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  constructor(props: IProps) {
    super(props);
    this.state = {
      hasError: false,
      error: '',
    };
  }

  componentDidCatch(error: any, info: any) {
    this.setState({
      error,
      info,
    });
    const { onError } = this.props;
    if (onError && typeof onError === 'function') {
      onError(error, info);
    }
  }

  render() {
    const { children, ...restProps } = this.props;
    const { hasError, error, info } = this.state;
    if (hasError) {
      return (
        // @ts-ignore
        (process.env.NODE_ENV === 'development' && (
          <DefaultFallbackComponent
            {...restProps}
            componentStack={info ? info.componentStack : ''}
            error={error}
          />
        )) || <Exception500 />
      );
    }
    return children;
  }
}

export default ErrorBoundary;
