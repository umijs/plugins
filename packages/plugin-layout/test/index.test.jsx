import getLayoutRenderConfig from '../src/layout/getLayoutRenderConfig';
import BlankLayout from '../src/layout/blankLayout';
import Logo from '../src/component/logo';
import { WithExceptionOpChildren } from '../src/component/Exception';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Service } from '@umijs/core';
import React from 'react';
import { join } from 'path';

const fixtures = join(__dirname, '..', 'fixtures');

if (!window.matchMedia) {
  Object.defineProperty(global.window, 'matchMedia', {
    writable: true,
    configurable: true,
    value: jest.fn((query) => ({
      matches: query.includes('max-width'),
      addListener: jest.fn(),
      removeListener: jest.fn(),
    })),
  });
}

afterEach(cleanup);

describe('getLayoutRenderConfig', () => {
  beforeEach(async () => {
    console.error = jest.fn();
    const service = new Service({
      cwd: fixtures,
      presets: ['@umijs/preset-built-in'],
      plugins: [require.resolve(join(fixtures, '..', 'src', 'index.ts'))],
    });
    await service.run({
      name: 'g',
      args: {
        _: ['g', 'tmp'],
      },
    });
  });

  it('getLayoutRenderConfig layout=false', () => {
    const layoutConfig = getLayoutRenderConfig(
      {
        layout: false,
      },
      false,
    );
    expect(layoutConfig.pure).toBeTruthy();
  });

  it('getLayoutRenderConfig hideMenu:true,hideFooter:true,hideNav:true', () => {
    const layoutConfig = getLayoutRenderConfig(
      {
        layout: { hideMenu: true, hideFooter: true, hideNav: true },
      },
      false,
    );
    expect(layoutConfig.menuRender).toBeFalsy();
    expect(layoutConfig.footerRender).toBeFalsy();
    expect(layoutConfig.headerRender).toBeFalsy();
  });

  it('getLayoutRenderConfig', () => {
    const layoutConfig = getLayoutRenderConfig({
      layout: {},
      hideFooter: true,
    });
    expect(layoutConfig.footerRender).toBeFalsy();
  });

  it('BlankLayout', () => {
    const { container } = render(<BlankLayout>Hello, World!</BlankLayout>);
    expect(container.firstChild).toMatchInlineSnapshot(`Hello, World!`);
  });

  it('ProLayout', () => {
    const Layout = require(join(
      fixtures,
      '.umi-test',
      'plugin-layout',
      'layout',
      'layout',
      'index.tsx',
    )).default;
    const { container } = render(
      <Layout location={{ pathname: '/' }}>Hello, World!</Layout>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('logo', () => {
    const { container } = render(<Logo />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('ProLayout rightContentRender', () => {
    const Layout = require(join(
      fixtures,
      '.umi-test',
      'plugin-layout',
      'layout',
      'layout',
      'index.tsx',
    )).default;
    const { container } = render(
      <Layout location={{ pathname: '/' }} rightContentRender={() => 'name'}>
        Hello, World!
      </Layout>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('ProLayout rightContentRender', () => {
    const Layout = require(join(
      fixtures,
      '.umi-test',
      'plugin-layout',
      'layout',
      'layout',
      'index.tsx',
    )).default;
    const { container } = render(
      <Layout
        location={{ pathname: '/' }}
        userConfig={{
          childrenRender: () => 'hello childrenRender',
        }}
        rightContentRender={() => 'name'}
      >
        Hello, World!
      </Layout>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('ProLayout patchMenus', () => {
    const Layout = require(join(
      fixtures,
      '.umi-test',
      'plugin-layout',
      'layout',
      'layout',
      'index.tsx',
    )).default;
    const { container } = render(
      <BrowserRouter>
        <Layout
          location={{ pathname: '/' }}
          userConfig={{
            patchMenus: () => [
              {
                name: 'qixian',
                path: '/',
              },
              {
                name: 'qixian2',
                path: '/qixian',
              },
            ],
          }}
          rightContentRender={() => 'name'}
        >
          Hello, World!
        </Layout>
      </BrowserRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('WithExceptionOpChildren 404', () => {
    const { container } = render(
      <WithExceptionOpChildren location={{ pathname: '/' }}>
        Hello, World!
      </WithExceptionOpChildren>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('WithExceptionOpChildren 403', () => {
    const { container } = render(
      <WithExceptionOpChildren
        currentPathConfig={{ unAccessible: true }}
        location={{ pathname: '/' }}
      >
        Hello, World!
      </WithExceptionOpChildren>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('WithExceptionOpChildren is ok', () => {
    const { container } = render(
      <WithExceptionOpChildren
        currentPathConfig={{ unAccessible: false }}
        location={{ pathname: '/' }}
      >
        Hello, World!
      </WithExceptionOpChildren>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
