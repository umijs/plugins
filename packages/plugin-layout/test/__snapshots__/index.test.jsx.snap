// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLayoutRenderConfig BlankLayout 2`] = `
<div
  class="ant-design-pro ant-pro-basicLayout screen-xl ant-pro-basicLayout-fix-siderbar ant-pro-basicLayout-side"
>
  <section
    class="ant-layout ant-layout-has-sider"
    style="min-height: 100%;"
  >
    <div
      style="width: 48px; overflow: hidden; flex: 0 0 48px; max-width: 48px; min-width: 48px; transition: background-color 0.3s, min-width 0.3s, max-width 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);"
    />
    <aside
      class="ant-layout-sider ant-layout-sider-dark ant-layout-sider-collapsed ant-layout-sider-below ant-pro-sider ant-pro-sider-fixed ant-pro-sider-layout-side"
      style="overflow: hidden; flex: 0 0 48px; max-width: 48px; min-width: 48px; width: 48px;"
    >
      <div
        class="ant-layout-sider-children"
      >
        <div
          class="ant-pro-sider-logo ant-pro-sider-collapsed"
          id="logo"
        >
          <a>
            <svg
              height="32"
              viewBox="0 0 200 200"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  id="linearGradient-1"
                  x1="62.102%"
                  x2="108.197%"
                  y1="0%"
                  y2="37.864%"
                >
                  <stop
                    offset="0%"
                    stop-color="#4285EB"
                  />
                  <stop
                    offset="100%"
                    stop-color="#2EC7FF"
                  />
                </lineargradient>
                <lineargradient
                  id="linearGradient-2"
                  x1="69.644%"
                  x2="54.043%"
                  y1="0%"
                  y2="108.457%"
                >
                  <stop
                    offset="0%"
                    stop-color="#29CDFF"
                  />
                  <stop
                    offset="37.86%"
                    stop-color="#148EFF"
                  />
                  <stop
                    offset="100%"
                    stop-color="#0A60FF"
                  />
                </lineargradient>
                <lineargradient
                  id="linearGradient-3"
                  x1="69.691%"
                  x2="16.723%"
                  y1="-12.974%"
                  y2="117.391%"
                >
                  <stop
                    offset="0%"
                    stop-color="#FA816E"
                  />
                  <stop
                    offset="41.473%"
                    stop-color="#F74A5C"
                  />
                  <stop
                    offset="100%"
                    stop-color="#F51D2C"
                  />
                </lineargradient>
                <lineargradient
                  id="linearGradient-4"
                  x1="68.128%"
                  x2="30.44%"
                  y1="-35.691%"
                  y2="114.943%"
                >
                  <stop
                    offset="0%"
                    stop-color="#FA8E7D"
                  />
                  <stop
                    offset="51.264%"
                    stop-color="#F74A5C"
                  />
                  <stop
                    offset="100%"
                    stop-color="#F51D2C"
                  />
                </lineargradient>
              </defs>
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="none"
                stroke-width="1"
              >
                <g
                  transform="translate(-20 -20)"
                >
                  <g
                    transform="translate(20 20)"
                  >
                    <g>
                      <g
                        fill-rule="nonzero"
                      >
                        <g>
                          <path
                            d="M91.588 4.177L4.18 91.513a11.981 11.981 0 000 16.974l87.408 87.336a12.005 12.005 0 0016.989 0l36.648-36.618c4.209-4.205 4.209-11.023 0-15.228-4.208-4.205-11.031-4.205-15.24 0l-27.783 27.76c-1.17 1.169-2.945 1.169-4.114 0l-69.802-69.744c-1.17-1.169-1.17-2.942 0-4.11l69.802-69.745c1.17-1.169 2.944-1.169 4.114 0l27.783 27.76c4.209 4.205 11.032 4.205 15.24 0 4.209-4.205 4.209-11.022 0-15.227L108.581 4.056c-4.719-4.594-12.312-4.557-16.993.12z"
                            fill="url(#linearGradient-1)"
                          />
                          <path
                            d="M91.588 4.177L4.18 91.513a11.981 11.981 0 000 16.974l87.408 87.336a12.005 12.005 0 0016.989 0l36.648-36.618c4.209-4.205 4.209-11.023 0-15.228-4.208-4.205-11.031-4.205-15.24 0l-27.783 27.76c-1.17 1.169-2.945 1.169-4.114 0l-69.802-69.744c-1.17-1.169-1.17-2.942 0-4.11l69.802-69.745c2.912-2.51 7.664-7.596 14.642-8.786 5.186-.883 10.855 1.062 17.009 5.837L108.58 4.056c-4.719-4.594-12.312-4.557-16.993.12z"
                            fill="url(#linearGradient-2)"
                          />
                        </g>
                        <path
                          d="M153.686 135.855c4.208 4.205 11.031 4.205 15.24 0l27.034-27.012c4.7-4.696 4.7-12.28 0-16.974l-27.27-27.15c-4.218-4.2-11.043-4.195-15.254.013-4.209 4.205-4.209 11.022 0 15.227l18.418 18.403c1.17 1.169 1.17 2.943 0 4.111l-18.168 18.154c-4.209 4.205-4.209 11.023 0 15.228z"
                          fill="url(#linearGradient-3)"
                        />
                      </g>
                      <ellipse
                        cx="100.519"
                        cy="100.437"
                        fill="url(#linearGradient-4)"
                        rx="23.6"
                        ry="23.581"
                      />
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </a>
        </div>
        <div
          style="flex: 1; overflow-y: auto; overflow-x: hidden;"
        />
        <div
          class="ant-pro-sider-links"
        >
          <ul
            class="ant-menu ant-menu-root ant-menu-vertical ant-menu-dark ant-pro-sider-link-menu ant-menu-inline-collapsed"
            data-menu-list="true"
            role="menu"
            tabindex="0"
          >
            <li
              class="ant-menu-item ant-menu-item-only-child ant-pro-sider-collapsed-button"
              data-menu-id="rc-menu-uuid-test-collapsed"
              role="menuitem"
              tabindex="-1"
            >
              <span
                class="ant-menu-title-content"
              >
                <span
                  aria-label="menu-unfold"
                  class="anticon anticon-menu-unfold"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="menu-unfold"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM142.4 642.1L298.7 519a8.84 8.84 0 000-13.9L142.4 381.9c-5.8-4.6-14.4-.5-14.4 6.9v246.3a8.9 8.9 0 0014.4 7z"
                    />
                  </svg>
                </span>
              </span>
            </li>
          </ul>
          <div
            aria-hidden="true"
            style="display: none;"
          />
        </div>
      </div>
    </aside>
    <div
      class="ant-layout"
      style="position: relative;"
    >
      <header
        class="ant-layout-header"
        style="height: 48px; line-height: 48px; background: transparent;"
      />
      <header
        class="ant-layout-header ant-pro-fixed-header ant-pro-header-dark"
        style="padding: 0px; height: 48px; line-height: 48px; width: 100%; z-index: 19; right: 0px;"
      >
        <div
          class="ant-pro-global-header ant-pro-global-header-layout-side"
        >
          <div
            style="flex: 1;"
          />
        </div>
      </header>
      <main
        class="ant-layout-content ant-pro-basicLayout-content ant-pro-basicLayout-has-header ant-pro-basicLayout-content-disable-margin"
      >
        Hello, World!
      </main>
    </div>
  </section>
</div>
`;
